pkg/controller/pxc/replication.go

```diff
@@ -86,6 +97,259 @@ func (r *ReconcilePerconaXtraDBCluster) removeOutdatedServices(cr *api.PerconaXt
+				db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, pod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
+				if err != nil {
+					return errors.Wrapf(err, "failed to connect to pod %s", pod.Name)
+				}
+				err = db.StopAllReplication()
+				db.Close()
+				if err != nil {
+					return errors.Wrapf(err, "stop replication on pod %s", pod.Name)
+				}
+				delete(pod.Labels, replicationPodLabel)
+				err = r.client.Update(context.TODO(), &pod)
+				if err != nil {
+					return errors.Wrap(err, "failed to remove primary label from secondary pod")
+				}
+			}
+		}
+	}
+
+	sysUsersSecretObj := corev1.Secret{}
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&sysUsersSecretObj,
+	)
+	if err != nil {
+		return errors.Wrap(err, "get secrets")
+	}
+
+	for _, channels := range cr.Spec.PXC.ReplicationChannels {
+		if channels.IsSource {
+			continue
+		}
+		err = manageReplicationChannel(r.log, primaryDB, channels, string(sysUsersSecretObj.Data["replication"]))
+		if err != nil {
+			return errors.Wrapf(err, "manage replication channel %s", channels.Name)
+		}
+	}
+
+	return nil
+}
+
+func removeOutdatedChannels(db queries.Database, current []api.ReplicationChannel) error {
+	channels, err := db.CurrentReplicationChannels()
+	if err != nil {
+		return errors.Wrap(err, "get current replication channels")
+	}
+
+	if len(channels) == 0 {
+		return nil
+	}
+
+	toRemove := make(map[string]struct{})
+	for _, v := range channels {
+		toRemove[v] = struct{}{}
+	}
+
+	for _, v := range current {
+		delete(toRemove, v.Name)
+	}
+
+	if len(toRemove) == 0 {
+		return nil
+	}
+
+	for channelToRemove := range toRemove {
+		err = db.StopReplication(channelToRemove)
+		if err != nil {
+			return errors.Wrapf(err, "stop replication for channel %s", channelToRemove)
+		}
+
+		srcList, err := db.ReplicationChannelSources(channelToRemove)
+		if err != nil {
+			return errors.Wrapf(err, "get src list for outdated channel %s", channelToRemove)
+		}
+		for _, v := range srcList {
+			err = db.DeleteReplicationSource(channelToRemove, v.Host, v.Port)
+			if err != nil {
+				return errors.Wrapf(err, "delete replication source for outdated channel %s", channelToRemove)
+			}
+		}
+	}
+	return nil
+}
+
+func manageReplicationChannel(log logr.Logger, primaryDB queries.Database, channel api.ReplicationChannel, replicaPW string) error {
+	currentSources, err := primaryDB.ReplicationChannelSources(channel.Name)
+	if err != nil && err != queries.ErrNotFound {
+		return errors.Wrapf(err, "get current replication sources for channel %s", channel.Name)
+	}
+
+	if err == queries.ErrNotFound {
+		for _, src := range channel.SourcesList {
+			err := primaryDB.AddReplicationSource(channel.Name, src.Host, src.Port, src.Weight)
+			if err != nil {
+				return errors.Wrapf(err, "add replication source for channel %s", channel.Name)
+			}
+		}
+	}
```

Comment:
why these lines needed, we will add sources anyway later

Reply:
87fe405